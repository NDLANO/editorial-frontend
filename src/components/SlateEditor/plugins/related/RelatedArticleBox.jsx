/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import { Transforms } from 'slate';
import { ReactEditor } from 'slate-react';
import PropTypes from 'prop-types';
import { uuid } from '@ndla/util';
import { injectT } from '@ndla/i18n';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { css } from '@emotion/core';
import { RelatedArticleList } from '@ndla/ui';
import { toggleRelatedArticles } from '@ndla/article-scripts';
import { convertFieldWithFallback } from '../../../../util/convertFieldWithFallback';
import { fetchDraft } from '../../../../modules/draft/draftApi';
import { queryResources } from '../../../../modules/taxonomy';
import { getLocale } from '../../../../modules/locale/locale';
import { EditorShape, AttributesShape } from '../../../../shapes';
import EditRelated from './EditRelated';
import handleError from '../../../../util/handleError';
import RelatedArticle from './RelatedArticle';
import { ARTICLE_EXTERNAL } from '../../../../constants';

const mapRelatedArticle = (article, resource) => ({
  ...article,
  resource,
  id: `${article.id}`,
  title: convertFieldWithFallback(article, 'title', article.title),
});

export class RelatedArticleBox extends React.Component {
  constructor() {
    super();
    this.state = { articles: [], editMode: false };
    this.fetchArticle = this.fetchArticle.bind(this);
    this.updateArticles = this.updateArticles.bind(this);
    this.insertExternal = this.insertExternal.bind(this);
    this.onInsertBlock = this.onInsertBlock.bind(this);
    this.openEditMode = this.openEditMode.bind(this);
    this.setNodeData = this.setNodeData.bind(this);
  }

  componentDidMount() {
    const {
      element: { data },
    } = this.props;
    if (data && data.nodes) {
      const articleNodes = data.nodes;
      this.fetchArticles(articleNodes).then(articles =>
        this.setState({ articles: articles.filter(a => !!a) }),
      );
    } else {
      this.setState({ editMode: true });
    }
  }

  componentDidUpdate() {
    // need to re-add eventhandler on button
    if (!this.state.editMode && this.state.articles.length > 2) {
      toggleRelatedArticles();
    }
  }

  onInsertBlock(newArticle) {
    if (!this.state.articles.find(it => it.id === parseInt(newArticle, 10))) {
      // get resource and add to state
      this.fetchArticle(newArticle).then(article => {
        if (article) {
          this.setState(oldState => ({
            articles: [...oldState.articles, article],
          }));
          this.setNodeData();
        }
      });
    }
  }

  setNodeData() {
    const { editor, element } = this.props;
    const { articles } = this.state;
    const path = ReactEditor.findPath(editor, element);
    Transforms.setNodes(
      editor,
      {
        data: {
          nodes: articles.map(article =>
            article.id === ARTICLE_EXTERNAL
              ? {
                  resource: 'related-content',
                  url: article.url,
                  title: article.title,
                }
              : { resource: 'related-content', 'article-id': article.id },
          ),
        },
      },
      { at: path, voids: true },
    );
  }

  structureExternal(url, title) {
    return {
      id: ARTICLE_EXTERNAL,
      tempId: uuid(),
      url,
      title,
      description: '',
    };
  }

  async fetchArticle(id) {
    try {
      const { locale } = this.props;
      const [article, resource] = await Promise.all([
        fetchDraft(id, locale),
        queryResources(id, locale),
      ]);
      if (article) {
        return mapRelatedArticle(article, resource);
      }
    } catch (error) {
      handleError(error);
    }
  }
  async fetchArticles(nodes) {
    return Promise.all(
      nodes.map(node => {
        const articleId = node['article-id'];
        if (articleId) {
          return this.fetchArticle(articleId);
        } else if (node.title) {
          return this.structureExternal(node.url, node.title);
        }
        return undefined;
      }),
    );
  }

  async insertExternal(url, title) {
    // await get description meta data
    this.setState(
      prevState => ({
        articles: [...prevState.articles, this.structureExternal(url, title)],
        editMode: false,
      }),
      this.setNodeData,
    );
  }

  updateArticles(newArticles) {
    this.setState({ articles: newArticles.filter(a => !!a) }, this.setNodeData);
  }

  openEditMode(e) {
    e.stopPropagation();
    this.setState({ editMode: true });
  }

  render() {
    const { attributes, onRemoveClick, locale, t, children } = this.props;
    const { editMode, articles } = this.state;

    if (editMode) {
      return (
        <EditRelated
          onRemoveClick={onRemoveClick}
          articles={articles}
          locale={locale}
          insertExternal={this.insertExternal}
          onInsertBlock={this.onInsertBlock}
          onExit={() => this.setState({ editMode: false })}
          updateArticles={this.updateArticles}
          {...attributes}
        />
      );
    }

    return (
      <div
        role="button"
        draggable
        contentEditable={false}
        tabIndex={0}
        data-testid="relatedWrapper"
        onClick={this.openEditMode}
        onKeyPress={this.openEditMode}
        css={css`
          & article > p {
            font-family: Source Sans Pro !important;
          }
        `}
        {...attributes}>
        <RelatedArticleList
          messages={{
            title: t('form.related.title'),
            showMore: t('form.related.showMore'),
            showLess: t('form.related.showLess'),
          }}>
          {articles.map((item, i) =>
            !item.id ? (
              t('form.content.relatedArticle.invalidArticle')
            ) : (
              <RelatedArticle key={uuid()} numberInList={i} item={item} />
            ),
          )}
        </RelatedArticleList>
        {children}
      </div>
    );
  }
}

RelatedArticleBox.propTypes = {
  attributes: AttributesShape,
  editor: EditorShape.isRequired,
  element: PropTypes.any,
  locale: PropTypes.string.isRequired,
  onRemoveClick: PropTypes.func,
};

const mapStateToProps = state => ({
  locale: getLocale(state),
});

export default compose(injectT, connect(mapStateToProps, null))(RelatedArticleBox);
