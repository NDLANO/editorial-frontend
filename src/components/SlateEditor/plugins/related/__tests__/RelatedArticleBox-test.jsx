/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import nock from 'nock';
import { createEditor } from 'slate';
import { render, fireEvent, cleanup, wait } from '@testing-library/react';
import { RelatedArticleBox } from '../RelatedArticleBox';
import IntlWrapper from '../../../../../util/__tests__/IntlWrapper';
import { TYPE_SECTION } from '../../section';
import { defaultRelatedBlock } from '..';

afterEach(cleanup);

const relatedArticleBlock = defaultRelatedBlock();

const wrapper = () => {
  const slate = [
    {
      type: TYPE_SECTION,
      children: [relatedArticleBlock],
    },
  ];
  const editor = createEditor();
  editor.children = slate;
  return render(
    <IntlWrapper>
      <RelatedArticleBox
        t={() => 'injected'}
        editor={editor}
        locale="nb"
        element={relatedArticleBlock}
      />
    </IntlWrapper>,
  );
};

jest.mock('slate-react');

test('it goes in and out of edit mode', async () => {
  nock('http://ndla-api')
    .get('/search-api/v1/search/editorial/?context-types=standard%2C%20topic-article&page=1&query=')
    .reply(200, {});
  const { getByTestId, container } = wrapper();

  fireEvent.click(getByTestId('showAddExternal'));
  expect(container.firstChild).toMatchSnapshot();

  const input = getByTestId('addExternalUrlInput');
  const inputTitle = getByTestId('addExternalTitleInput');
  fireEvent.change(input, { target: { value: 'https://www.vg.no' } });
  fireEvent.change(inputTitle, { target: { value: 'Verdens gang' } });

  fireEvent.click(getByTestId('taxonomyLightboxButton'));
  await wait();

  expect(container.firstChild).toMatchSnapshot();
});
