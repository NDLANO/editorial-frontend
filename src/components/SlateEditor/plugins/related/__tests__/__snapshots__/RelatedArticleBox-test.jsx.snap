// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it goes in and out of edit mode 1`] = `


.emotion-55 article>p {
  font-family: Source Sans Pro!important;
}

.emotion-19 {
  position: absolute;
  top: 0.1rem;
  right: 0.2rem;
  color: #d1372e;
}

.emotion-19:hover,
.emotion-19:focus {
  color: #7d211c;
}

.emotion-27 {
  height: 50px;
  width: 50px;
}

.emotion-26 {
  height: 24px;
  width: 24px;
  margin-right: 7px;
}



.emotion-44,
.emotion-44:hover {
  border-radius: 5px;
  background-color: white;
  margin-top: 26px;
  padding: 3px 52px;
}

.emotion-53 {
  position: absolute;
  width: 100%;
}

.emotion-0 {
  position: fixed;
  top: 0;
  left: 0;
  height: 100vh;
  width: 100vw;
  background: rgba(0,0,0,0.3);
  -webkit-animation-duration: 400ms;
  animation-duration: 400ms;
  -webkit-animation-name: fadeIn;
  animation-name: fadeIn;
}

.emotion-22 {
  position: relative;
  border: 2px solid #a5bcd3;
  padding: 52px;
  padding-top: 0;
  background-color: #fff;
  z-index: 1;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  z-index: 1;
}

.emotion-4 figure {
  position: static !important;
}

.emotion-2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-17 {
  position: relative;
  -webkit-flex: 1 0 50%;
  -ms-flex: 1 0 50%;
  flex: 1 0 50%;
  max-width: 600px;
}

.emotion-17 > article {
  max-width: 100%;
}

.emotion-11 {
  margin-top: 6.5px;
}

.emotion-11:first-of-type {
  margin-top: 0;
}

.emotion-9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-9 label {
  width: 104px;
  max-width: 104px;
  padding: 20px 13px 13px 0;
  text-transform: uppercase;
  font-weight: 600;
  font-size: 14;
  font-size: 0.7777777777777778rem;
  line-height: 1.1;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  background: #f8f8f8;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  border: 1px solid #eff0f2;
  -webkit-transition: border-color 100ms ease;
  transition: border-color 100ms ease;
  border-radius: 4px;
  min-height: 52px;
  padding-right: 13px;
}

.emotion-8 input,
.emotion-8 textarea {
  width: inherit;
  font-weight: 400;
  color: #444;
  font-size: 18;
  font-size: 1rem;
  line-height: 1.3;
  background: none;
  border: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.emotion-8 input:focus,
.emotion-8 textarea:focus {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  outline: none;
}

.emotion-8 input {
  padding: 6.5px 13px;
}

.emotion-8 textarea {
  padding: 0 13px;
  height: 20px;
  margin: 14px 0;
  resize: none;
}

.emotion-8 .c-icon {
  width: 24px;
  height: 24px;
}

.emotion-13 {
  margin: 10px 0;
}

.emotion-15 {
  display: inline-block;
  color: #ffffff;
  background-color: #20588f;
  border: 2px solid #20588f;
  border-radius: 4px;
  padding: 4px 13px;
  outline-width: 0;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 16px;
  font-size: 0.8888888888888888rem;
  line-height: 1.625;
  font-family: 'Source Sans Pro',Helvetica,Arial,STKaiti,'华文楷体',KaiTi,SimKai,'楷体',KaiU,DFKai-SB,'標楷體',SongTi,'宋体',sans-serif;
  font-weight: 700;
  -webkit-transition: all .2s cubic-bezier(0.17,0.04,0.03,0.94);
  transition: all .2s cubic-bezier(0.17,0.04,0.03,0.94);
  box-shadow: none;
  text-align: center;
}

.emotion-15:hover,
.emotion-15:focus {
  color: white;
  background-color: #184673;
  border: 2px solid #184673;
}

.emotion-15[disabled] {
  color: #777777;
  background-color: #e6e6e6;
  border-color: transparent;
  cursor: not-allowed;
}

.emotion-15:focus {
  box-shadow: 0 0 2px #20588f;
}

.emotion-20 {
  display: inline-block;
  color: #ffffff;
  background-color: #20588f;
  border: 2px solid #20588f;
  border-radius: 4px;
  padding: 4px 13px;
  outline-width: 0;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 16px;
  font-size: 0.8888888888888888rem;
  line-height: 1.625;
  font-family: 'Source Sans Pro',Helvetica,Arial,STKaiti,'华文楷体',KaiTi,SimKai,'楷体',KaiU,DFKai-SB,'標楷體',SongTi,'宋体',sans-serif;
  font-weight: 700;
  -webkit-transition: all .2s cubic-bezier(0.17,0.04,0.03,0.94);
  transition: all .2s cubic-bezier(0.17,0.04,0.03,0.94);
  box-shadow: none;
  text-align: center;
  -webkit-transition: background-color none;
  transition: background-color none;
  padding: 0;
  border-radius: 0;
  color: inherit;
  font-size: inherit;
  background-color: transparent;
  box-shadow: none;
  border: none;
  font-weight: 400;
}

.emotion-20:hover,
.emotion-20:focus {
  color: white;
  background-color: #184673;
  border: 2px solid #184673;
}

.emotion-20[disabled] {
  color: #777777;
  background-color: #e6e6e6;
  border-color: transparent;
  cursor: not-allowed;
}

.emotion-20:focus {
  box-shadow: 0 0 2px #20588f;
}

.emotion-20:hover,
.emotion-20:active,
.emotion-20:disabled,
.emotion-20:focus {
  box-shadow: none;
  color: #20588f;
  background-color: transparent;
  border: none;
}

.emotion-20:focus {
  outline-width: medium;
}

.emotion-6 {
  border: calc(26px / 6.5) solid rgba(0,0,0,0.1);
  border-bottom-color: #20588f;
  border-radius: 50%;
  -webkit-animation: spinnerAnimation 0.7s linear infinite;
  animation: spinnerAnimation 0.7s linear infinite;
  height: 26px;
  width: 26px;
  display: block;
  margin: 0;
}

.emotion-51 {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0,0,0,0.8);
  z-index: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-49 {
  background-color: #f8f8f8;
  box-shadow: 0 0 2px 0 rgba(115,115,115,0.5);
  width: 620px;
  border-radius: 5px;
  position: absolute;
  top: 5%;
  z-index: 2;
  max-height: 90vh;
  overflow: auto;
}

.emotion-30 {
  background: linear-gradient(180deg,#5d97a9,#508a9c);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1.2rem;
  padding-left: 13px;
  height: 50px;
  color: #f8f8f8;
}

.emotion-47 {
  padding: 2em;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.emotion-42 {
  width: 100%;
}

.emotion-42 > * {
  width: 100%;
}

.emotion-42 form {
  background-color: white;
}

<div>
  <div
    autocapitalize="false"
    autocorrect="false"
    contenteditable="true"
    data-gramm="false"
    data-slate-editor="true"
    data-slate-node="value"
    role="textbox"
    spellcheck="false"
    style="position: relative; outline: none; white-space: pre-wrap; word-wrap: break-word;"
    zindex="-1"
  >
    <div
      data-slate-node="element"
      style="position: relative;"
    >
      <div
        data-slate-node="element"
        style="position: relative;"
      >
        <span
          data-slate-node="text"
        >
          <span
            data-slate-leaf="true"
          >
            <span
              data-slate-length="0"
              data-slate-zero-width="n"
            >
              ﻿
              <br />
            </span>
          </span>
        </span>
      </div>
    </div>
  </div>
  <div
    class="emotion-53 emotion-54"
    contenteditable="false"
  >
    <div
      aria-hidden="true"
      class="emotion-0 emotion-1"
    />
    <div
      class="emotion-22 emotion-23"
    >
      <h1
        class="c-section-heading c-related-articles__component-title"
      >
        Relatert innhold
      </h1>
      <p>
        Dra artikkel for å endre rekkefølge
      </p>
      <div
        class="emotion-4 emotion-5"
      >
        <div
          class="drop-zone emotion-2 emotion-3"
        />
      </div>
      <div
        class="emotion-17 emotion-18"
        data-cy="styled-article-modal"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="downshift-0-label"
          role="combobox"
          style="position: relative;"
        >
          <div
            class="emotion-11 emotion-12"
          >
            <div
              class="emotion-9 emotion-10"
            >
              <form
                class="emotion-8"
              >
                <input
                  aria-autocomplete="list"
                  aria-labelledby="downshift-0-label"
                  autocomplete="off"
                  data-testid="dropdownInput"
                  id="downshift-0-input"
                  placeholder="Søk på tittel"
                  value=""
                />
                <div
                  class="emotion-6 emotion-7"
                />
              </form>
            </div>
          </div>
        </div>
        <div
          class="emotion-13 emotion-14"
        >
          Eller
        </div>
        <button
          class="emotion-15 emotion-16"
          data-testid="showAddExternal"
          type="button"
        >
          Legg til ekstern artikkel
        </button>
      </div>
      <button
        class="emotion-19 emotion-20 emotion-16"
        contenteditable="false"
        data-cy="close-related-button"
        type="button"
      >
        <svg
          class="c-icon"
          data-license="Apache License 2.0"
          data-source="Material Design"
          fill="currentColor"
          height="1em"
          preserveAspectRatio="xMidYMid meet"
          style="vertical-align: middle;"
          viewBox="0 0 24 24"
          width="1em"
        >
          <title>
            DeleteForever
          </title>
          <g>
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </g>
        </svg>
      </button>
    </div>
    <div
      class="emotion-51 emotion-52"
    >
      <div
        aria-hidden="true"
        class="emotion-0 emotion-1"
      />
      <div
        class="emotion-49 emotion-50"
      >
        <div
          class="emotion-30 emotion-31"
        >
          Skriv inn tittel og url på ekstern artikkel
          <button
            class="emotion-27 emotion-20 emotion-16"
            type="button"
          >
            <svg
              class="c-icon emotion-26"
              data-license="Apache License 2.0"
              data-source="Material Design"
              fill="currentColor"
              height="1em"
              preserveAspectRatio="xMidYMid meet"
              style="vertical-align: middle;"
              viewBox="0 0 24 24"
              width="1em"
            >
              <title>
                Cross
              </title>
              <g>
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </g>
            </svg>
          </button>
        </div>
        <div
          class="emotion-47 emotion-48"
        >
          <div
            class="emotion-42 emotion-43"
          >
            <div
              class="emotion-11 emotion-12"
            >
              <div
                class="emotion-9 emotion-10"
              >
                <div
                  class="emotion-8"
                >
                  <input
                    data-testid="addExternalTitleInput"
                    placeholder="Tittel"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="emotion-11 emotion-12"
            >
              <div
                class="emotion-9 emotion-10"
              >
                <div
                  class="emotion-8"
                >
                  <input
                    data-testid="addExternalUrlInput"
                    placeholder="https://example.com"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <button
            class="emotion-44 emotion-20 emotion-16"
            data-testid="taxonomyLightboxButton"
            type="button"
          >
            Lagre
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="emotion-55"
    contenteditable="false"
    data-testid="relatedWrapper"
    draggable="true"
    role="button"
    tabindex="0"
  >
    <section
      class="c-related-articles"
    >
      <h1
        class="c-section-heading c-related-articles__component-title"
      >
        Relatert innhold
      </h1>
      <div
        class="c-related-articles__articles"
      />
    </section>
  </div>
</div>
`;

exports[`it goes in and out of edit mode 2`] = `


.emotion-24 article>p {
  font-family: Source Sans Pro!important;
}

.emotion-17 {
  position: absolute;
  top: 0.1rem;
  right: 0.2rem;
  color: #d1372e;
}

.emotion-17:hover,
.emotion-17:focus {
  color: #7d211c;
}

.emotion-22 {
  position: absolute;
  width: 100%;
}

.emotion-0 {
  position: fixed;
  top: 0;
  left: 0;
  height: 100vh;
  width: 100vw;
  background: rgba(0,0,0,0.3);
  -webkit-animation-duration: 400ms;
  animation-duration: 400ms;
  -webkit-animation-name: fadeIn;
  animation-name: fadeIn;
}

.emotion-20 {
  position: relative;
  border: 2px solid #a5bcd3;
  padding: 52px;
  padding-top: 0;
  background-color: #fff;
  z-index: 1;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  z-index: 1;
}

.emotion-4 figure {
  position: static !important;
}

.emotion-2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-15 {
  position: relative;
  -webkit-flex: 1 0 50%;
  -ms-flex: 1 0 50%;
  flex: 1 0 50%;
  max-width: 600px;
}

.emotion-15 > article {
  max-width: 100%;
}

.emotion-9 {
  margin-top: 6.5px;
}

.emotion-9:first-of-type {
  margin-top: 0;
}

.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-7 label {
  width: 104px;
  max-width: 104px;
  padding: 20px 13px 13px 0;
  text-transform: uppercase;
  font-weight: 600;
  font-size: 14;
  font-size: 0.7777777777777778rem;
  line-height: 1.1;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  background: #f8f8f8;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  border: 1px solid #eff0f2;
  -webkit-transition: border-color 100ms ease;
  transition: border-color 100ms ease;
  border-radius: 4px;
  min-height: 52px;
  padding-right: 13px;
}

.emotion-6 input,
.emotion-6 textarea {
  width: inherit;
  font-weight: 400;
  color: #444;
  font-size: 18;
  font-size: 1rem;
  line-height: 1.3;
  background: none;
  border: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.emotion-6 input:focus,
.emotion-6 textarea:focus {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  outline: none;
}

.emotion-6 input {
  padding: 6.5px 13px;
}

.emotion-6 textarea {
  padding: 0 13px;
  height: 20px;
  margin: 14px 0;
  resize: none;
}

.emotion-6 .c-icon {
  width: 24px;
  height: 24px;
}

.emotion-11 {
  margin: 10px 0;
}

.emotion-13 {
  display: inline-block;
  color: #ffffff;
  background-color: #20588f;
  border: 2px solid #20588f;
  border-radius: 4px;
  padding: 4px 13px;
  outline-width: 0;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 16px;
  font-size: 0.8888888888888888rem;
  line-height: 1.625;
  font-family: 'Source Sans Pro',Helvetica,Arial,STKaiti,'华文楷体',KaiTi,SimKai,'楷体',KaiU,DFKai-SB,'標楷體',SongTi,'宋体',sans-serif;
  font-weight: 700;
  -webkit-transition: all .2s cubic-bezier(0.17,0.04,0.03,0.94);
  transition: all .2s cubic-bezier(0.17,0.04,0.03,0.94);
  box-shadow: none;
  text-align: center;
}

.emotion-13:hover,
.emotion-13:focus {
  color: white;
  background-color: #184673;
  border: 2px solid #184673;
}

.emotion-13[disabled] {
  color: #777777;
  background-color: #e6e6e6;
  border-color: transparent;
  cursor: not-allowed;
}

.emotion-13:focus {
  box-shadow: 0 0 2px #20588f;
}

.emotion-18 {
  display: inline-block;
  color: #ffffff;
  background-color: #20588f;
  border: 2px solid #20588f;
  border-radius: 4px;
  padding: 4px 13px;
  outline-width: 0;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 16px;
  font-size: 0.8888888888888888rem;
  line-height: 1.625;
  font-family: 'Source Sans Pro',Helvetica,Arial,STKaiti,'华文楷体',KaiTi,SimKai,'楷体',KaiU,DFKai-SB,'標楷體',SongTi,'宋体',sans-serif;
  font-weight: 700;
  -webkit-transition: all .2s cubic-bezier(0.17,0.04,0.03,0.94);
  transition: all .2s cubic-bezier(0.17,0.04,0.03,0.94);
  box-shadow: none;
  text-align: center;
  -webkit-transition: background-color none;
  transition: background-color none;
  padding: 0;
  border-radius: 0;
  color: inherit;
  font-size: inherit;
  background-color: transparent;
  box-shadow: none;
  border: none;
  font-weight: 400;
}

.emotion-18:hover,
.emotion-18:focus {
  color: white;
  background-color: #184673;
  border: 2px solid #184673;
}

.emotion-18[disabled] {
  color: #777777;
  background-color: #e6e6e6;
  border-color: transparent;
  cursor: not-allowed;
}

.emotion-18:focus {
  box-shadow: 0 0 2px #20588f;
}

.emotion-18:hover,
.emotion-18:active,
.emotion-18:disabled,
.emotion-18:focus {
  box-shadow: none;
  color: #20588f;
  background-color: transparent;
  border: none;
}

.emotion-18:focus {
  outline-width: medium;
}

<div>
  <div
    autocapitalize="false"
    autocorrect="false"
    contenteditable="true"
    data-gramm="false"
    data-slate-editor="true"
    data-slate-node="value"
    role="textbox"
    spellcheck="false"
    style="position: relative; outline: none; white-space: pre-wrap; word-wrap: break-word;"
    zindex="-1"
  >
    <div
      data-slate-node="element"
      style="position: relative;"
    >
      <div
        data-slate-node="element"
        style="position: relative;"
      >
        <span
          data-slate-node="text"
        >
          <span
            data-slate-leaf="true"
          >
            <span
              data-slate-length="0"
              data-slate-zero-width="n"
            >
              ﻿
              <br />
            </span>
          </span>
        </span>
      </div>
    </div>
  </div>
  <div
    class="emotion-22 emotion-23"
    contenteditable="false"
  >
    <div
      aria-hidden="true"
      class="emotion-0 emotion-1"
    />
    <div
      class="emotion-20 emotion-21"
    >
      <h1
        class="c-section-heading c-related-articles__component-title"
      >
        Relatert innhold
      </h1>
      <p>
        Dra artikkel for å endre rekkefølge
      </p>
      <div
        class="emotion-4 emotion-5"
      >
        <div
          class="drop-zone emotion-2 emotion-3"
        />
      </div>
      <div
        class="emotion-15 emotion-16"
        data-cy="styled-article-modal"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="downshift-0-label"
          role="combobox"
          style="position: relative;"
        >
          <div
            class="emotion-9 emotion-10"
          >
            <div
              class="emotion-7 emotion-8"
            >
              <form
                class="emotion-6"
              >
                <input
                  aria-autocomplete="list"
                  aria-labelledby="downshift-0-label"
                  autocomplete="off"
                  data-testid="dropdownInput"
                  id="downshift-0-input"
                  placeholder="Søk på tittel"
                  value=""
                />
                <svg
                  class="c-icon"
                  data-license="Apache License 2.0"
                  data-source="Material Design"
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style="vertical-align: middle;"
                  viewBox="0 0 48 48"
                  width="1em"
                >
                  <title>
                    Search
                  </title>
                  <g>
                    <path
                      d="M31 28h-1.59l-.55-.55C30.82 25.18 32 22.23 32 19c0-7.18-5.82-13-13-13S6 11.82 6 19s5.82 13 13 13c3.23 0 6.18-1.18 8.45-3.13l.55.55V31l10 9.98L40.98 38 31 28zm-12 0c-4.97 0-9-4.03-9-9s4.03-9 9-9 9 4.03 9 9-4.03 9-9 9z"
                    />
                  </g>
                </svg>
              </form>
            </div>
          </div>
        </div>
        <div
          class="emotion-11 emotion-12"
        >
          Eller
        </div>
        <button
          class="emotion-13 emotion-14"
          data-testid="showAddExternal"
          type="button"
        >
          Legg til ekstern artikkel
        </button>
      </div>
      <button
        class="emotion-17 emotion-18 emotion-14"
        contenteditable="false"
        data-cy="close-related-button"
        type="button"
      >
        <svg
          class="c-icon"
          data-license="Apache License 2.0"
          data-source="Material Design"
          fill="currentColor"
          height="1em"
          preserveAspectRatio="xMidYMid meet"
          style="vertical-align: middle;"
          viewBox="0 0 24 24"
          width="1em"
        >
          <title>
            DeleteForever
          </title>
          <g>
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </g>
        </svg>
      </button>
    </div>
  </div>
  <div
    class="emotion-24"
    contenteditable="false"
    data-testid="relatedWrapper"
    draggable="true"
    role="button"
    tabindex="0"
  >
    <section
      class="c-related-articles"
    >
      <h1
        class="c-section-heading c-related-articles__component-title"
      >
        Relatert innhold
      </h1>
      <div
        class="c-related-articles__articles"
      />
    </section>
  </div>
</div>
`;
